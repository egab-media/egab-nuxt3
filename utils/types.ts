import { UserRecord } from 'firebase-admin/lib/auth/user-record'

export interface EgabUser extends Partial<UserRecord> {
  username?: string
}
