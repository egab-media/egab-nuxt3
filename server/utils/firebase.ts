import { initializeApp, applicationDefault } from 'firebase-admin/app'
import { getAuth } from 'firebase-admin/auth'

export const adminApp = initializeApp({
  credential: applicationDefault(),
  projectId: 'egab-staging'
  // databaseURL: process.env.FIREBASE_DATABASE_URL
})

export const adminAuth = getAuth()
