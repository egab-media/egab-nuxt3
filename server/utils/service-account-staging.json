{
  "type": "service_account",
  "project_id": "egab-staging",
  "private_key_id": "b3639900bbb0f5e5c9bcaf28aa1d14e5a7dcd0b4",
  "private_key": "-----BEGIN PRIVATE KEY-----\nMIIEvQIBADANBgkqhkiG9w0BAQEFAASCBKcwggSjAgEAAoIBAQCj2U7laodej+T9\nkhA9XEMgD0XID10l7T5m+l8L9G8+cYkpj+0GADVdAIdWeaez4+IZQ3IRTQxzoK2v\nxVQzpG/CWKI19zQZFHwIt83EftmYG/Fwr11xvNT7Y9orfozWsdcXQ/E5SUiEe8FU\nAlwl9wVyWE2XVkmS9i1cS+s35CCiFf6zcKivput2M/m5vJQTMlX7L+8XXsRCBG6e\nrxazj6GKXIQyHpAIo8E6tRRSZCYhAG3wInwL/MQa9iQKEI/LiX7itNSpZLuG8nrj\nUBv7gNTjCk8P2093riQ5Is2krEqDaDMP+muIDSUF54WePe48E49trYK+e1lZD2JW\nSdYd4Ia9AgMBAAECggEAILcLi0AdflKR4NVwu6e32WdjQgJbUW7Q1mPoZW0tZxTp\nU6yioV+aSBMKcshApWj7ZJr0ZCLzuQScmWt3mAYdsHrdDJRqyRUvfNgiVDkVvkra\nButk54hSKE20KpoOifGhZkpRGGSa/6TsIZeFCamMY1vhRjXa4+aWgQbBlPvOTj5u\n2tyoFgHg5UGwsdykft2+Kztt8/IJqQNSEueWIiDNrYXgf8skd4HlrcwCTURH3Cn/\nJilsRArI6dR7jajylOmq7udI54kuXZ2PynbS7o29XE8iAm8jlgsnn0hLpDc21ZXK\nwHi0xZfOWjUoP6h5vWkGeRaO442FsC3Yw+cHQsX0iwKBgQDh0Uin8kSS5cNq8HTO\nlEDvQuEZmOSQUPpR0HXo2WLwg4oJ6PblJxUshcHhJgVInHmzR2XvKwYWwSH21wxD\nXjm4vMnqS0ErpE1MlbxeQSIyblPvmDo31KbWEoy4HkeHv98MuBNPmkAqs/SQoTsQ\naGLae1YJ5oFFEwHxxuaOuMDv4wKBgQC5v6pTvD2J8zH8Ak/Bl764kJ8FjfjRE2JD\nu5h7jgi8wimamf+5JlrrYTbFUW6pFBPJYK6m2aVSl8n083bbxcOD+QPLX2YrG3Me\njcXf/wFd2uS2g6/11kpTukOV8MOQsdGCU5H7oh4C8AhozISxXGpUM8WMHJaaaGVS\nFbm/Oecw3wKBgDtS4NSGN++jftgcPYN7vSdeGYbA/OLQ80IFEsTOP0jSNbRuyFXh\nE5XzucBtXZv6x5oKmsPfOKVHraWMT0NvgJiAuEbd6EwPyOLiv26/i71V8KUUB89d\nBUtd6iwAgGgwRuevgBs8ebj+N7S7NqoT0kx1dF/7d2BvIuaiizVZYg5jAoGARn8Y\nvOj/jiCvws55yKa36sJwqCIG4gryS3SiXZo9dK5YLeXuqS2GajkeGWtbjW4iSFaa\nUpbd6OfU7VONs8Ft/+N9bi1J0Gh9m4YBKI9WzHEsRpzaJnprATX7OqgDASKLumPP\nTupra4AAIQ4l+k5xyH1Rp2iePAINGkdGlO378Y8CgYEArBS781ooTe3zk3/sQ991\nU7WQOOj2GrMWXJePOiX6FgHXyYhH2fn3icl5c2G3PGDvPArwGzefbuFpYdfCWZOn\n6tllw+x+MjfKU4JIMcLmDU9FkUk2jjHWaxfNHmZA8ZoqouJ2lFp7G+jdGkDptR3L\nUwWHlDjBrnM/KM0XUuIa66A=\n-----END PRIVATE KEY-----\n",
  "client_email": "firebase-adminsdk-uviw7@egab-staging.iam.gserviceaccount.com",
  "client_id": "103454563138954494611",
  "auth_uri": "https://accounts.google.com/o/oauth2/auth",
  "token_uri": "https://oauth2.googleapis.com/token",
  "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
  "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/firebase-adminsdk-uviw7%40egab-staging.iam.gserviceaccount.com",
  "universe_domain": "googleapis.com"
}
