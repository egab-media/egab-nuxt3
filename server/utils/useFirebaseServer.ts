import { cert, getApps, initializeApp } from 'firebase-admin/app'
import { getAuth } from 'firebase-admin/auth'
// @ts-ignore
import serviceAccountJson from './service-account-staging.json'

export default function useFirebaseServer() {
  const app = getApps()[0] ?? initializeApp({
    credential: cert(serviceAccountJson)
  })

  const auth = getAuth(app)

  return { auth }
}
