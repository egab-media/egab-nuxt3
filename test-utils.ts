import { vi } from 'vitest'
import en from '@/locales/en-US.json'
import ar from '@/locales/ar-SA.json'
import { createTestingPinia } from '@pinia/testing'
import { createVuetify } from 'vuetify'
import * as vueTestUtils from '@vue/test-utils'
import { createApp } from 'vue'
import { createI18n } from 'vue-i18n'

export const addVuetify = (context: any) => {
    context.vuetifyInstance = createVuetify({});
    context.vueTestUtils.config.global.plugins = [context.vuetifyInstance];
    context.vueTestUtils.config.global.CSS = { supports: () => false };
}

// export const addVuex = (context: { vuex: any; vue: { use: (arg0: any) => void } }) => {
//     context.vuex = require('vuex')
//     context.vue.use(context.vuex)
// }

export const addPinia = (context: any) => {
    context.vueTestUtils.config.global.plugins.push(createTestingPinia())
}

export const addI18n = (context: any): any => {
    // eslint-disable-next-line new-cap
    context.i18nInstance = createI18n({
        // ...options,
        messages: { en, ar },
        fallbackLocale: 'en',
        legacy: true,
        globalInjection: true,
    })
    context.vueTestUtils.config.global.plugins.push(context.i18nInstance)
}

export const addFilter = (name: any, lambda: any): any => {
    return (context: { vue: any[] }) => context.vue.filter(name, lambda)
}

export const compositeConfiguration = (...configs: any[]): any => {
    return (context: any) => configs.forEach(config => config(context))
}

export const bootstrapVueContext = (configureContext: (arg0: {}) => any) => {
    const context: any = {}

    const teardownVueContext = () => {
        vi.unmock('vue')
        Object.keys(context).forEach(key => delete context[key])
        vi.resetModules()
    }

    context.vueTestUtils = vueTestUtils
    context.vue = createApp({})

    vi.doMock('vue', () => context.vue)

    configureContext && configureContext(context)

    return {
        teardownVueContext,
        ...context
    }
}

export const shallowMount = (component: any, context: any) => {
    return context.vueTestUtils.shallowMount(component, {
        localVue: context.localVue,
        vuetify: context.vuetifyInstance,
        i18n: context.i18nInstance,
        store: context.store,
        router: context.routerInstance,
        propsData: context.propsData,
        stubs: context.stubs,
        mocks: context.mocks,
        slots: context.slots,
        attachTo: context.attachTo
    })
}

export const mountWrapper = (component: any, context: any) => {
    return context.vueTestUtils.mount(component, {
        localVue: context.localVue,
        vuetify: context.vuetifyInstance,
        i18n: context.i18nInstance,
        store: context.store,
        router: context.routerInstance,
        propsData: context.propsData,
        stubs: context.stubs,
        mocks: context.mocks,
        slots: context.slots,
        attachTo: context.attachTo
    })
}
