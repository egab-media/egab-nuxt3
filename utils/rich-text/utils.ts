export const getUnitWithPxAsDefault = (value?: string | number) => {
  if (!value) { return value }

  const _value = isNumber(value) ? String(value) : value

  const num = parseInt(_value, 10)
  const unit = _value.slice(num.toString().length)

  return num + (unit || 'px')
}

export function clamp(val: number, min: number, max: number) {
  if (val < min) { return min }
  if (val > max) { return max }
  return val
}

/**
 * @method isNumber
 * Checks if a value is a number
 * @param value - Value to be checked
 * @returns A boolean indicating whether the value is a number or not.
 */
export const isNumber = (value: unknown): value is number => typeof value === 'number'
/**
 *  @method isString
 *  Checks if a value is a string.
 * @param value - The value to be checked.
 * @returns A boolean indicating whether the value is a string or not.
 */
export const isString = (value: unknown): value is string => typeof value === 'string'
/**
 * @method isBoolean
 * Checks if a value is a boolean.
 * @param value - The value to be checked.
 * @returns A boolean indicating whether the value is a boolean or not.
 */
export const isBoolean = (value: unknown): value is boolean => typeof value === 'boolean'
/**
 * @method isFunction
 * Checks if a value is a function.
 * @param value - The value to check.
 * @returns True if the value is a function, false otherwise.
 */
export const isFunction = (value: unknown): value is Function => typeof value === 'function'
