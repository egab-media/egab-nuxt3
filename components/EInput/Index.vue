<template>
  <v-col :cols="cols">
    <v-text-field
      v-model="reactiveValue"
      v-bind="$props"
      :ref="id"
      :id="id"
      :append-inner-icon="type === 'password' ? showPass ? mdiEye : mdiEyeOff : appendIcon"
      density="compact"
      :type="showPass ? 'text' : type"
      flat
      :title="id"
      variant="outlined"
      :rules="handleRules()"
      @keyup="type === 'password' ? initProgress(inputRef, rules) : false"
      @click:appendInner="showPass = !showPass"
    >
      <template
          v-for="(_, inputSlot) in $slots"
          #[inputSlot]="slotScope"
      >
        <slot :name="inputSlot" v-bind="slotScope" />
      </template>

      <template v-slot:label="{label, props}">
        <span v-if="rules.includes('required')" v-text="'* '" data-test="input-asterisk" class="red--text" />
        <span v-text="label" data-test="input-label" />
      </template>

      <template v-if="type === 'password'" #loader>
        <v-progress-linear
          v-for="(_rule, index) in rules"
          :key="index"
          :model-value="progress"
          :color="progress === 100 ? 'primary' : progress < 50 ? 'error' : progress < 70 ? 'orange' : 'warning'"
          :absolute="true"
          height="5"
          style="margin-top: 5px;"
        />
      </template>
    </v-text-field>
  </v-col>
</template>

<script lang="ts">
import * as ruleSpecs from '~/utils/allRules'

export default defineComponent({
  name: "EInput",
  emits: ['update:modelValue'],
  props: {
    cols: {
      type: [String, Number],
      default: 12
    },
    id: {
      type: String,
      default: 'test'
    },
    label: {
      type: String,
      default: null
    },
    hint: { type: String, default: undefined },
    persistentHint: { type: Boolean, default: false },
    modelValue: {
      default: ''
    },
    rules: {
      type: Array as () => Array<string>,
      default: () => []
    },
    type: {
      type: String,
      default: 'text'
    },
    loading: {
      type: Boolean,
      default: false
    },
    appendIcon: {
      type: String,
      default: ''
    },
    counter: { type: Boolean, default: false },
    counterValue: { type: Function, default: undefined }
  },
  data: (vm) => ({
    inputRef: null,
    showPass: false,
    // progress: 0,
    reservedRules: {
      required: (val: string) => ruleSpecs.required(val, vm.$t('auth.form.validation.required') as string),
      blank: (val: string) => ruleSpecs.blank(val, 'cannot be blank') as string,
      email: (val: string) => ruleSpecs.email(val, vm.$t('auth.form.validation.email') as string),
      alpha: (val: string) => ruleSpecs.alpha(val, vm.$t('auth.form.validation.alpha') as string),
      titleMinChars: (val: string) => ruleSpecs.titleMinChars(val, vm.$t('auth.form.validation.titleMinChars') as string),
      fullNameMinChars: (val: string) => ruleSpecs.fullNameMinChars(val, vm.$t('auth.form.validation.fullNameMinChars') as string),
      hasLowercase: (val: string) => ruleSpecs.hasLowercase(val, vm.$t('auth.form.validation.hasLowercase') as string),
      hasUppercase: (val: string) => ruleSpecs.hasUppercase(val, vm.$t('auth.form.validation.hasUppercase') as string),
      hasNumber: (val: string) => ruleSpecs.hasNumber(val, vm.$t('auth.form.validation.hasNumber') as string),
      hasSpecial: (val: string) => ruleSpecs.hasSpecial(val, vm.$t('auth.form.validation.hasSpecial') as string),
      minchars: (val: string) => ruleSpecs.minchars(val, vm.$t('auth.form.validation.minchars') as string),
      url: (val: string) => ruleSpecs.url(val)
    } as any
  }),

  computed: {
    reactiveValue: {
      get(): any {
        return this.modelValue
      },
      set(newValue: any) {
        this.$emit('update:modelValue', newValue)
      }
    }
  },

  mounted() {
    // Cache the input ref based on the passed id prop for handling validation and progress
    if (this.$refs && this.$refs[this.id]) {
      (this as any).inputRef = this.$refs[this.id]
    }
  },

  methods: {
    /**
     * @method handleRules
     * This takes the string values of the passed 'rules' prop
     * and executes the reserved rules stored in @/components/EInput/allRules.ts
     * ---
     * This is engineered for 4 reasons:
     * 1. To ensure having singularity for common validation rules all over the software
     * 2. To avoid multiple imports for the same rules in every parent component
     * 3. To isolate the rules as a business logic for unit testing (later)
     * 4. To easily maintain the rules from @/components/EInput/allRules.ts file
     * ---
     * @returns Array<Function>
     */
    handleRules(): Array<any> {
      const finalRules: any = []
      const reservedRules = this.reservedRules

      this.rules.forEach((key: string | any) => {
        if (typeof key !== 'string') {
          finalRules.push(key)
        }

        if (key in reservedRules) {
          finalRules.push(reservedRules[key])
        }
      })
      return finalRules
    },

    /**
     * DO NOT DELETE!
     * @method getProgress
     * Uses a service worker to calculate the progress for passing password requirements
     * This is engineered to ensure that the is no stress on the main browser thread
     * So it is preferable to assign this to another browser thread to avoid performance issues
     * @returns {number} progress
     */
    // async getProgress() {
    //   if (process.client) {
    //     let errs: any[] = []
    //     if (this.inputRef) {
    //       const cached = await (this.inputRef as any).validate()
    //       errs = toRaw(cached)
    //     }
    //
    //     const worker = new Worker('/js/password-progress.worker.js')
    //     // Listening
    //     worker.addEventListener('message', (event: { data: number }) => {
    //       this.progress = event.data
    //       worker.terminate()
    //     })
    //
    //     // initiating
    //     worker.postMessage({
    //       rules: this.rules,
    //       errs
    //     })
    //     return this.progress
    //   }
    // }
  }
})
</script>

<script lang="ts" setup>
import {VCol} from "vuetify/components/VGrid"
import {VTextField} from 'vuetify/components/VTextField';
import {VProgressLinear} from "vuetify/components/VProgressLinear";
import { mdiEye, mdiEyeOff } from '@mdi/js'
import {useGetProgress} from '@/composables/input-progress'

const { getProgress } = useGetProgress()
const progress = ref(0)
const initProgress = (inputRef: any, rules: string[]) => getProgress(inputRef, rules, toRef(progress))
defineExpose({ progress, getProgress })
</script>

<style scoped>

</style>
