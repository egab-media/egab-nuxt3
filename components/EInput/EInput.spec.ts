import {flushPromises, VueWrapper} from '@vue/test-utils'
import EInput from './Index.vue'
import { it, expect, describe, afterEach, beforeEach, vi } from 'vitest'
import {
    addI18n,
    addVuetify,
    bootstrapVueContext,
    compositeConfiguration,
    mountWrapper
} from '~/test-utils'
import {ComponentPublicInstance} from "@vue/runtime-core";
import * as ruleSpecs from "~/utils/allRules";

type MyComponentProps = any
type MyComponentVariables = {  }
let wrapper: VueWrapper<ComponentPublicInstance<any, any>>
let vueContext: any

const findAsterisk = () => wrapper.find('[data-test="input-asterisk"]')
const findLabelWrapper = () => wrapper.find('.v-label')
const findLabel = () => wrapper.find('[data-test="input-label"]')

describe('Global EInput', () => {
    vueContext = bootstrapVueContext(compositeConfiguration(addVuetify, addI18n))
    vueContext.mocks = {$worker: {}}
    vueContext.propsData = {id: 'vi'}
    // vueContext.stubs = ['v-progress-linear']
    beforeEach(() => {
        /**
         * SECTION: only for vue-phone-input plugin
         * mockIntersectionObserver is used to support polyfill
         */
        // const mockIntersectionObserver = vi.fn()
        // mockIntersectionObserver.mockReturnValue({
        //     observe: () => null,
        //     unobserve: () => null,
        //     disconnect: () => null
        // })
        // window.IntersectionObserver = mockIntersectionObserver

        wrapper = mountWrapper(EInput, vueContext)
    })

    afterEach(() => {
        vueContext.teardownVueContext()
    })

    describe('rendering component', () => {
        it('should load the component', async () => {
            expect(wrapper.vm).toBeTruthy()
        })
    })

    describe('Label', () => {
        it('should render asterisk', async () => {
            expect(findAsterisk().exists()).toBe(false)
            await wrapper.setProps({
                rules: ['required']
            })
            // Must research asterisk again in the new DOM
            expect(findAsterisk().exists()).toBe(true)
            expect(findAsterisk().text()).toEqual('*')
            expect(findAsterisk().attributes('class')).toEqual('red--text')
        })

        it.todo('should render label', async () => {
            const labelWrapper = findLabelWrapper()
            expect(labelWrapper.text()).toBe('vi')

            const labelText = findLabel()

            expect(labelText.exists()).toBeTruthy()
            expect(labelText.text()).toEqual('vi')
            expect(labelText.attributes('class')).toContain('text-caption text-capitalize font-weight-bold')

            await wrapper.setProps({
                rules: ['required']
            })
            expect(labelWrapper.text().replace(/\s+/, '')).toBe('*vi')
        })
    })

    // adds 20% branching cover
    describe('password', () => {
        it('should render pass', async function () {
            const input = wrapper.find('input')
            expect(input.attributes('type')).toEqual('text')
            await wrapper.setProps({ type: 'password', label: 'password' })
            expect(input.attributes('type')).toEqual('password')
            await wrapper.setData({
                showPass: true
            })
            expect(wrapper.find('input').attributes('type')).toEqual('text')
            expect(wrapper.find('label').text()).toBe('password')
        });

        it('should render icons', async function () {
            let iconWrapper;
            iconWrapper = wrapper.find('.v-field__append-inner')
            expect(iconWrapper.exists()).toBeFalsy()

            await wrapper.setProps({ type: 'password' })
            expect(wrapper.vm.appendIcon).toEqual('')
            iconWrapper = wrapper.find('.v-field__append-inner')
            expect(iconWrapper.exists()).toBeTruthy()
            await iconWrapper.find('i').trigger('click')
            expect(wrapper.vm.showPass).toBe(true)
            await iconWrapper.find('i').trigger('click')
            expect(wrapper.vm.showPass).toBe(false)
        });
    })

    describe('Rules', () => {
        // covers 2%
        it('should check required', async function () {
            await wrapper.setProps({ rules: ['required'] })
            expect(wrapper.find('.v-input--error').exists()).toBeFalsy()
            const input = wrapper.find('input')
            await input.trigger('focus')
            await input.trigger('blur')
            expect(wrapper.find('.v-input--error').exists()).toBeTruthy()
            expect(wrapper.find('.v-input__details').text()).toEqual('* required')
        });

        it('should check blank', async function () {
            await wrapper.setProps({ rules: ['blank'] })
            const input = wrapper.find('input')
            await input.trigger('focus')
            await input.trigger('blur')
            expect(wrapper.find('.v-input__details').text()).toEqual('cannot be blank')
        });

        it('should check email', async function () {
            await wrapper.setProps({ rules: ['email'] })
            const input = wrapper.find('input')
            await input.trigger('focus')
            await input.trigger('blur')
            expect(wrapper.find('.v-input__details').text()).toEqual('* Invalid e-mail.')
        });

        it('should check alpha', async function () {
            await wrapper.setProps({ rules: ['alpha'] })
            const input = wrapper.find('input')
            await input.trigger('focus')
            await input.trigger('blur')
            expect(wrapper.find('.v-input__details').text()).toEqual('Only letters are allowed')
        });

        it('should check titleMinChars', async function () {
            await wrapper.setProps({ rules: ['titleMinChars'] })
            const input = wrapper.find('input')
            await input.trigger('focus')
            await input.trigger('blur')
            expect(wrapper.find('.v-input__details').text()).toEqual('* title must have at least 5 words')
        });

        it('should check fullNameMinChars', async function () {
            await wrapper.setProps({ rules: ['fullNameMinChars'] })
            const input = wrapper.find('input')
            await input.trigger('focus')
            await input.trigger('blur')
            expect(wrapper.find('.v-input__details').text()).toEqual('* First and Last name required')
        });

        it('should check hasLowercase', async function () {
            await wrapper.setProps({ rules: ['hasLowercase'] })
            const input = wrapper.find('input')
            await input.trigger('focus')
            await input.trigger('blur')
            expect(wrapper.find('.v-input__details').text()).toEqual('* password must have at least 1 lowercase character')
        });

        it('should check hasUppercase', async function () {
            await wrapper.setProps({ rules: ['hasUppercase'] })
            const input = wrapper.find('input')
            await input.trigger('focus')
            await input.trigger('blur')
            expect(wrapper.find('.v-input__details').text()).toEqual('* password must have at least 1 uppercase character')
        });

        it('should check hasNumber', async function () {
            await wrapper.setProps({ rules: ['hasNumber'] })
            const input = wrapper.find('input')
            await input.trigger('focus')
            await input.trigger('blur')
            expect(wrapper.find('.v-input__details').text()).toEqual('* password must have at least one number')
        });

        it('should check hasSpecial', async function () {
            await wrapper.setProps({ rules: ['hasSpecial'] })
            const input = wrapper.find('input')
            await input.trigger('focus')
            await input.trigger('blur')
            expect(wrapper.find('.v-input__details').text()).toEqual('* password must have at least one special character (\'"!@#$*;&:?-+)')
        });

        it('should check minchars', async function () {
            await wrapper.setProps({ rules: ['minchars'] })
            const input = wrapper.find('input')
            await input.trigger('focus')
            await input.trigger('blur')
            expect(wrapper.find('.v-input__details').text()).toEqual('* password must have at least 6 characters')
        });

        it('should check url', async function () {
            await wrapper.setProps({ rules: ['url'] })
            const input = wrapper.find('input')
            await input.trigger('focus')
            await input.trigger('blur')
            expect(wrapper.find('.v-input__details').text()).toEqual('* Wrong facebook url format')
        });

        it.todo('should check computed rules', async function () {
            await wrapper.setProps({ rules: ['required'] })
            console.log(wrapper.vm.handleRules())
            expect(typeof wrapper.vm.rules[0]).toBe('string')
            expect(wrapper.vm.handleRules()[0]).toContain(function required(val: string) {})
        });
    })

    describe('keyup event', () => {
        it('should test keyup', async function () {
            const spy = vi.spyOn(wrapper.vm, 'initProgress')
            const input = wrapper.find('input')
            await input.trigger('keyup')
            expect(spy).not.toHaveBeenCalled()

            await wrapper.setProps({ type: 'password', rules: ['required', 'alpha'] })
            await input.setValue('ee')
            await input.trigger('keyup')
            expect(spy).toHaveBeenCalled()
        });
    })

    describe('progress bar', () => {
        // Eliminates 14% of branching
        it('should render the progress correctly', async function () {
            await wrapper.setProps({ type: 'password', rules: ['required', 'hasSpecial'] })
            wrapper.vm.progress = 100
            await nextTick()

            const progressBar = wrapper.find('.v-progress-linear')
            const innerBar = progressBar.find('.v-progress-linear__determinate')
            expect(progressBar.attributes('aria-valuenow')).toEqual('100')
            expect(innerBar.attributes('class')).toContain('bg-primary')

            wrapper.vm.progress = 50
            await nextTick()
            expect(progressBar.attributes('aria-valuenow')).toEqual('50')
            expect(innerBar.attributes('class')).toContain('bg-orange')

            wrapper.vm.progress = 80
            await nextTick()
            expect(progressBar.attributes('aria-valuenow')).toEqual('80')
            expect(innerBar.attributes('class')).toContain('bg-warning')
        });
    })
})
