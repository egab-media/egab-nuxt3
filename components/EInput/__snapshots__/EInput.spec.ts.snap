// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Global EInput > Label > should render label 1`] = `
"<div class=\\"v-col v-col-12\\">
  <div class=\\"v-input v-input--horizontal v-input--center-affix v-input--density-compact v-locale--is-ltr v-text-field\\">
    <!---->
    <div class=\\"v-input__control\\">
      <div class=\\"v-field v-field--center-affix v-field--flat v-field--variant-outlined v-theme--light v-locale--is-ltr\\" role=\\"textbox\\">
        <div class=\\"v-field__overlay\\"></div>
        <div class=\\"v-field__loader\\">
          <div class=\\"v-progress-linear v-theme--light v-locale--is-ltr\\" style=\\"top: 0px; height: 0px; --v-progress-linear-height: 2px; left: 50%; transform: translateX(-50%);\\" role=\\"progressbar\\" aria-hidden=\\"true\\" aria-valuemin=\\"0\\" aria-valuemax=\\"100\\">
            <!---->
            <div class=\\"v-progress-linear__background\\" style=\\"width: 100%;\\"></div>
            <transition-stub name=\\"fade-transition\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
              <div class=\\"v-progress-linear__indeterminate\\">
                <div class=\\"v-progress-linear__indeterminate long\\"></div>
                <div class=\\"v-progress-linear__indeterminate short\\"></div>
              </div>
            </transition-stub>
            <!---->
          </div>
        </div>
        <!---->
        <div class=\\"v-field__field\\" data-no-activator=\\"\\">
          <!----><label class=\\"v-label v-field-label\\" for=\\"vi\\">
            <!---->
            <!--v-if--><span data-test=\\"input-label\\"></span>
          </label>
          <!---->
          <div class=\\"v-field__input\\" data-no-activator=\\"\\"><input size=\\"1\\" type=\\"text\\" id=\\"vi\\" aria-describedby=\\"vi-messages\\" cols=\\"12\\" title=\\"vi\\"></div>
          <!---->
        </div>
        <!---->
        <!---->
        <div class=\\"v-field__outline\\">
          <div class=\\"v-field__outline__start\\"></div>
          <div class=\\"v-field__outline__notch\\"><label class=\\"v-label v-field-label v-field-label--floating\\" aria-hidden=\\"true\\" for=\\"vi\\">
              <!---->
              <!--v-if--><span data-test=\\"input-label\\"></span>
            </label></div>
          <div class=\\"v-field__outline__end\\"></div>
          <!---->
        </div>
      </div>
    </div>
    <!---->
    <div class=\\"v-input__details\\">
      <transition-group-stub name=\\"slide-y-transition\\" tag=\\"div\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\" class=\\"v-messages\\" role=\\"alert\\" aria-live=\\"polite\\" id=\\"vi-messages\\">
        <!---->
      </transition-group-stub>
      <!---->
    </div>
  </div>
</div>"
`;
